File: go_algorithms/flatten_array.js
function flatten(currentArr, depth = 1) {
    let temp = []
    
    if (depth === 0) {
        return currentArr
    }

    for (i = 0; i < currentArr.length; i++) {
        if (typeof currentArr[i] === "number") {
            temp.push(currentArr[i])
        } else {
            currentArr[i].forEach(elem => {
                temp.push(elem)
            })
        }
    }

    return flatten(temp, depth - 1)
}

console.log(flatten([1, 2, 3, [4, 5, [6, 7]], [8, 9]], 2))

File: go_algorithms/flatten_array.go
/*

Count depth
	If at the appropriate depth, return current slice
	Else, return outermost array plus flattened inner arrays
5 layer array, flatten to 1

The optional parameter specifies how many levels to flatten, not up to what flatness

*/

package main

import "fmt"

type mixed interface{}

func flatten(current []interface{}, depth int) []interface{} {
	result := []interface{}{}

	if depth == 0 {
		return current
	} else {
		for _, elem := range current {
			switch t := elem.(type) {
			case int:
				result = append(result, elem)
			case mixed:
				for i := 0; i < ; i++ {
				}
			default:
				fmt.Println(t)
			}
		}
	}

	return result

	//return flatten(current, depth-1)
}

func main() {
	fmt.Println(flatten([]interface{}{1, 2, []interface{}{3, 4, []interface{}{5, 6}}}, 1))
}
